sdk/iot/hub/src/az_iot_hub_client_sas.c

```diff
@@ -14,110 +14,165 @@
-  remainder = az_span_copy(remainder, sr_string);
-  remainder = az_span_copy_u8(remainder, EQUAL_SIGN);
-  remainder = az_span_copy(remainder, iothub_fqdn);
-  remainder = az_span_copy(remainder, devices_string);
-  remainder = az_span_copy(remainder, device_id);
+  mqtt_password_span = az_span_copy(mqtt_password_span, sr_string);
+  mqtt_password_span = az_span_copy_u8(mqtt_password_span, EQUAL_SIGN);
+  mqtt_password_span = az_span_copy(mqtt_password_span, client->_internal.iot_hub_hostname);
+
+  // Device ID
+  mqtt_password_span = az_span_copy(mqtt_password_span, devices_string);
+  mqtt_password_span = az_span_copy(mqtt_password_span, client->_internal.device_id);
+
+  // Module ID
+  if (az_span_size(client->_internal.options.module_id) > 0)
+  {
+    mqtt_password_span = az_span_copy(mqtt_password_span, modules_string);
+    mqtt_password_span = az_span_copy(mqtt_password_span, client->_internal.options.module_id);
+  }
 
   // Signature
-  remainder = az_span_copy_u8(remainder, AMPERSAND);
-  remainder = az_span_copy(remainder, sig_string);
-  remainder = az_span_copy_u8(remainder, EQUAL_SIGN);
-  remainder = az_span_copy(remainder, signature);
+  mqtt_password_span = az_span_copy_u8(mqtt_password_span, AMPERSAND);
+  mqtt_password_span = az_span_copy(mqtt_password_span, sig_string);
+  mqtt_password_span = az_span_copy_u8(mqtt_password_span, EQUAL_SIGN);
+  mqtt_password_span = az_span_copy(mqtt_password_span, base64_hmac_sha256_signature);
 
   // Expiration
-  remainder = az_span_copy_u8(remainder, AMPERSAND);
-  remainder = az_span_copy(remainder, se_string);
-  remainder = az_span_copy_u8(remainder, EQUAL_SIGN);
-  AZ_RETURN_IF_FAILED(az_span_i32toa(remainder, expiry_time_secs, &remainder));
+  mqtt_password_span = az_span_copy_u8(mqtt_password_span, AMPERSAND);
+  mqtt_password_span = az_span_copy(mqtt_password_span, se_string);
+  mqtt_password_span = az_span_copy_u8(mqtt_password_span, EQUAL_SIGN);
+  AZ_RETURN_IF_FAILED(az_span_u32toa(mqtt_password_span, token_expiration_epoch_time, &mqtt_password_span));
 
   if (az_span_size(key_name) > 0)
   {
-    az_span skn_string = AZ_SPAN_FROM_STR(SAS_TOKEN_SKN);
-    required_length = az_span_size(skn_string) + az_span_size(key_name) + 2;
-
-    AZ_RETURN_IF_NOT_ENOUGH_SIZE(remainder, required_length);
-
     // Key Name
-    remainder = az_span_copy_u8(remainder, AMPERSAND);
-    remainder = az_span_copy(remainder, skn_string);
-    remainder = az_span_copy_u8(remainder, EQUAL_SIGN);
-    remainder = az_span_copy(remainder, key_name);
+    mqtt_password_span = az_span_copy_u8(mqtt_password_span, AMPERSAND);
+    mqtt_password_span = az_span_copy(mqtt_password_span, skn_string);
+    mqtt_password_span = az_span_copy_u8(mqtt_password_span, EQUAL_SIGN);
+    mqtt_password_span = az_span_copy(mqtt_password_span, key_name);
+  }
+
+  if (out_mqtt_password_length != NULL)
+  {
+    *out_mqtt_password_length = (mqtt_password_size - (size_t)az_span_size(mqtt_password_span));
```

Comment:
>*out_mqtt_password_length = (mqtt_password_size - (size_t)az_span_size(mqtt_password_span)); [](start = 4, length = 92)

out_mqtt_password_length should tell you how much buffer you filled or left. Right?

*out_mqtt_password_length = (size_t)az_span_size(mqtt_password_span));

Reply:
Actually no, that is not how the new az_span_copy functions work. They return the rest of the input az_span (the part that has free space). So if you just do az_span_size of mqtt_password_span, it gives you the remaining free space of the original buffer. We can talk in person if you prefer.

Reply:
Thanks for the clarification. 

---
In reply to: [414746132](https://github.com/Azure/azure-sdk-for-c/pull/568#discussion_r414746132) [](ancestors = 414746132)